### PR TITLE
Add support for additional APIs, test coverage

### PR DESCRIPTION
**Pull request by kujenga**
This PR adds support for a few more APIs from https://docs.helium.com/api and adds test coverage for them (and others) too. I'm new-ish to Rust to happy to iterate on any feedback!

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -133,9 +164,13 @@ impl Client {
 
     pub(crate) fn fetch<T: DeserializeOwned>(&self, path: &str) -> error::Result<T> {
         let request_url = format!("{}{}", self.base_url, path);
-        let response = self.client.get(&request_url).send()?.error_for_status()?;
+        return Ok(self.fetch_data(request_url)?.data);
+    }
+
+    pub(crate) fn fetch_data<T: DeserializeOwned>(&self, url: String) -> error::Result<Data<T>> {
```
Comment (madninja) [inline src/lib.rs:170]: Make url a `&str`?

File: src/lib.rs [hunk 2]
```diff
@@ -159,6 +194,11 @@ impl Client {
         self.fetch::<Account>(&format!("/accounts/{}", address))
     }
 
+    /// Get wallet information for the richest accounts
+    pub fn get_accounts_richest(&self, limit: Option<u64>) -> error::Result<Vec<Account>> {
```
Comment (madninja) [inline src/lib.rs:198]: a u64 for limit is perhaps a bit overlay optimistic at how large pages can be.. u16 or u32 at best

File: src/lib.rs [hunk 3]
```diff
@@ -175,6 +215,56 @@ impl Client {
...
+    pub fn get_hotspot_rewards(
+        &self,
+        address: &str,
+        min_time: Option<&str>,
+        max_time: Option<&str>,
+        cursor: Option<&str>,
+    ) -> error::Result<(Vec<Reward>, Option<String>)> {
+        let base = Url::parse(&format!("{}/", self.base_url)).unwrap();
+        let mut request_url = base.join(&format!("hotspots/{}/rewards", address)).unwrap();
+        for (k, v) in [
```
Comment (madninja) [inline src/lib.rs:228]: Try using `Url::parse_with_params` after creating a vec of options that have values.

File: src/lib.rs [hunk 4]
```diff
@@ -175,6 +215,56 @@ impl Client {
...
+            ("min_time", min_time),
+            ("max_time", max_time),
+            ("cursor", cursor),
+        ]
+        .iter()
+        {
+            if v.is_some() {
+                request_url.query_pairs_mut().append_pair(k, v.unwrap());
```
Comment (madninja) [inline src/lib.rs:236]: See above.. I think parse_with_params may be a cleaner solution. If not, at least hang on to the results of query_pairs_mut so you can call append_pair without re-requesting the mutable query pairs every time?

File: src/lib.rs [hunk 5]
```diff
@@ -175,6 +215,56 @@ impl Client {
...
+            }
+        }
+
+        let data = self.fetch_data(request_url.to_string())?;
+        Ok((data.data, data.cursor))
+    }
+
+    pub fn get_oracle_price_current(&self) -> error::Result<OraclePrice> {
```
Comment (madninja) [inline src/lib.rs:244]: match API name -> `get_oracle_prices_current`

**Review by madninja — CHANGES_REQUESTED** [review]
Thanks for your PR!! 

A few changes requested but this is looking pretty good!

File: src/lib.rs (hunk 3 above)
Comment (kujenga) [reply src/lib.rs]: Nice, I've added a `build_url` private helper to encapsulate that and the "optional param" logic, swapping out the previous implementation.

File: src/lib.rs (hunk 4 above)
Comment (kujenga) [reply src/lib.rs]: This now uses the `build_url` helper

**Comment by kujenga** [comment]
I believe I've addressed all the feedback, thanks! Not sure what's happening with the build, I'm not able to see what the failures are (there's no dropdown for output), and I'm developing on mac and see no issues locally. Maybe it just needs to be re-run?

Screen Shot 2021-02-23 at 19 11 37

File: src/lib.rs [hunk 6]
```diff
@@ -87,6 +88,35 @@ pub struct Hotspot {
...
+pub struct OraclePrice {
+    /// The price submitted by the oracle in 1/100,000,000 USD.
+    pub price: u64,
+    /// The block height when the reward was received.
+    pub block: u64,
+}
+
+impl OraclePrice {
+    /// The price submitted by the oracle in USD.
+    pub fn price_usd(&self) -> f64 {
```
Comment (madninja) [inline src/lib.rs:115]: rename to `to_usd()`

File: src/lib.rs [hunk 7]
```diff
@@ -131,11 +162,27 @@ impl Client {
         Self { base_url, client }
     }
 
+    pub(crate) fn build_url(&self, path: &str, params: &[(&str, Option<&str>)]) -> reqwest::Url {
+        // Create a filtered set of string params that were specified.
+        let params = params.iter().filter_map(|p| {
+            if p.1.is_none() {
```
Comment (madninja) [inline src/lib.rs:168]: this is neat.. There has to be a more idiomatic way to do this though?

Comment (kujenga) [reply src/lib.rs:187]: Sure, happy to iterate here, though some guidance on what direction is most idiomatic would be most welcome! Two possibilities come to mind:

Clarify flow a bit by splitting out the filter/map stages:

```rust
        let params = params
            .iter()
            .filter(|p| p.1.is_some())
            .map(|p| (p.0, p.1.unwrap()));
```

Or this is how I would write it if I was trying to write idiomatic Go code:

```rust
        let mut actual_params = Vec::new();
        for p in params {
            if p.1.is_some() {
                actual_params.push((p.0, p.1.unwrap()));
            }
        }
```

If something else entirely comes to mind let me know.

Comment (kujenga) [reply src/lib.rs]: I've pushed the first one as a starting point

**Comment by kujenga** [comment]
Rebased, and attempted to address the comments 👍

**Comment by madninja** [comment]
I figured it may be easier to show what I'm working on: https://github.com/helium/helium-api-rs/pull/24 and see how we can get your additions into this newer proposed approach?

**Comment by kujenga** [comment]
Ah, neat! Yeah happy to re-work this on top of those changes. Once that's in I can take another pass here.

**Comment by madninja** [comment]
I'm going to close this since the new main branch I believe has what was in this branch. Thanks for triggering me to finish it of @kujenga !

**Pull request closed without merging**